Sources/TokamakCore/Shapes/Path.swift

```diff
@@ -139,6 +143,27 @@ public struct Path: Equatable, LosslessStringConvertible {
     case let .roundedRect(fixedRoundedRect): return fixedRoundedRect.rect
     case let .stroked(strokedPath): return strokedPath.path.boundingRect
     case let .trimmed(trimmedPath): return trimmedPath.path.boundingRect
+    case let .path(pathBox):
+      // Note: Copied from TokamakStaticHTML/Shapes/Path.swift
```

Comment:
Is this copied from that file or actually moved? Is the overall duplication reduced then?

Reply:
I made a single change. I'll take a look and see if I can get a better understanding of the existing logic and try to reuse the same logic both places.

Reply:
I _think_ that there was an error in the previous sizing and that it's fixed in the `Path.boundingRect` function, but I am not 100% certain.

The previous `elementsSize` in StaticHTML/Shapes/Path.swift had:
```
    return CGSize(width: abs(maxX - min(0, minX)), height: abs(maxY - min(0, minY)))
```
but I replaced that with
```
    return CGSize(width: maxX - minX, height: maxY - minY)
```
because otherwise the size depends on the origin, which I guess it shouldn't - unless the usage of the size needs this special behavior.
But the similar special casing was previously not applied for .rect storage type - here the size was just the rect's size (no matter the origin) - so the change seems more consistent.

@carson-katri Do you remember a use case for the special case for the previous element path sizing?

Reply:
BTW, all our renderer-independent code in TokamakCore is easy to test and there's already a test target for it, which runs only on macOS for now. I'll make it run on Ubuntu too in a future PR. Thus if you think you've found a bug in something like this, please provide a test case. Then we can test the new code against it, if that makes sense 🙂

Reply:
I don't remember, sorry 😞

Reply:
Hi @MaxDesiatov ,
The code was in the TokamakStaticHTML target, and I guess the best way to test would be to get a visual representation of various shapes with positive and negative origins - and compare them to SwiftUI.
I will have a look at this when I get a chance.

Reply:
Ok, I think I can gather the point of the old code. I cannot figure out what's best/correct though.

The whole thing has to do with sizing of `Shape`. In SwiftUI a view is being proposed a size from it's parent - a `Shape` always accepts the proposed size.
If 'no size' is proposed (e.g. by using .fixedSize() modifier), a `Shape` uses the fallback of a 10 by 10 frame.

In other words - SwiftUI has no 'intrinsic' size for a `Shape`.
In `TokamakCore` there is added to the `Shape` a concept of fixed and flexible sizing. This concept does not appear to be part of SwiftUI (where the sizing is always determined as a function of the proposed size).
And in `TokamakStaticHTML` an intrinsic size for a `Shape` (or rather, a `Path`) is being calculated - and for the .fixed size concept, this is where the `size` calculation was used.

So previously it appeared that the calculation was basically:
The bounding rect of the `Path` 'unioned' with .zero (or something to that effect). Meaning that a bounding rect with origin 10, 10 and size 10, 10 would generate the size 20, 20.
If the path was entirely in the negative axises I think that it would be cropped away since the calculated size is always in positive.

Currently the boundingRect size is being returned, which I guess is not much better.

But the question is: what is actually the best thing to do?

SwiftUI has a different way of sizing and if there is such a thing as an intrinsic size, it appears to be 10 by 10 (which is the size of a `Shape` when applying .fixedSize()).

Reply:
Another comment about the `_Sizing` concept:
Right now built-in shapes like `Rectangle` are .flexible in size while custom Shapes are .fixed.
A custom shape may as well be flexible - it all depends on whether or not the `func path(in rect:)` returns a path that is relative to the rect or not...

I don't know if the _Sizing concept could be made obsolete somehow by doing something else entirely...

Reply:
I consider bringing the layout stuff in a separate PR from https://github.com/objcio/S01E235-swiftui-layout-explained-layout-priorities, would that be related in any way?

Reply:
I appear to get ok results by removing the `Path._Sizing` enum entirely and choosing the code paths that corresponded to the previous 'flexible' sizing...

That removes the need for the fixed size entirely - and I get quite similar results with SwiftUI and Tokamak(DOM) drawing shapes with fixed sizes - both when I create them using a 'Shape' or a 'Path' directly.

<img width="504" alt="Skærmbillede 2021-01-12 kl  10 52 19" src="https://user-images.githubusercontent.com/1384769/104299850-fe7fa200-54c5-11eb-836b-49d44fc31832.png">
<img width="1792" alt="Skærmbillede 2021-01-12 kl  10 52 07" src="https://user-images.githubusercontent.com/1384769/104299856-017a9280-54c6-11eb-8384-d1100b67201b.png">

I am still uncertain if I may be lacking some context, but at least I suspect that this could be a way to go about it...

Reply:
If fixing this requires some substantial changes, I suggest doing it in a separate PR to make changes more granular and easier to track and review.

Reply:
Excellent. I’ll leave the sizing as it is for now. I don’t think it’s substantially more correct or incorrect than the previous version.

And regarding your question above which I just saw now:
Yes, my knowledge about the layout system indeed comes from the reverse engineering work of objc.io Swift Talks. 

I really hope to get time to look into that for GTK.
I believe that the full layout system requires you to be able to measure view sizes before they are rendered - I think that can be done in gtk, but I am not familiar with how it might be done in HTML/JavaScript although there appears to be some hacks available.